addons/survey/controllers/main.py

```diff
@@ -732,35 +727,113 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
+
+        The matching user_input_lines are all the lines tied to the user inputs which respect
+        the survey base domain and which have lines matching all the filters.
+        For example, with the filter 'Where do you live?|Brussels', we need to display ALL the lines
+        of the survey user inputs which have answered 'Brussels' to this question.
+
+        :return (recordset, List[dict]): all matching user input lines, each search filter data
+        """
+        user_input_line_subdomains = []
         search_filters = []
-        line_filter_domain, line_choices = [], []
-        for data in post.get('filters', '').split('|'):
-            try:
-                row_id, answer_id = (int(item) for item in data.split(','))
-            except:
-                pass
-            else:
-                if row_id and answer_id:
-                    line_filter_domain = expression.AND([
-                        ['&', ('matrix_row_id', '=', row_id), ('suggested_answer_id', '=', answer_id)],
-                        line_filter_domain
-                    ])
-                    answers = request.env['survey.question.answer'].browse([row_id, answer_id])
-                elif answer_id:
-                    line_choices.append(answer_id)
-                    answers = request.env['survey.question.answer'].browse([answer_id])
-                if answer_id:
-                    question_id = answers[0].matrix_question_id or answers[0].question_id
-                    search_filters.append({
-                        'row_id': row_id,
-                        'answer_id': answer_id,
-                        'question': question_id.title,
-                        'answers': '%s%s' % (answers[0].value, ': %s' % answers[1].value if len(answers) > 1 else '')
-                    })
-        if line_choices:
-            line_filter_domain = expression.AND([[('suggested_answer_id', 'in', line_choices)], line_filter_domain])
 
-        user_input_domain = self._get_user_input_domain(survey, line_filter_domain, **post)
-        user_input_lines = request.env['survey.user_input'].sudo().search(user_input_domain).mapped('user_input_line_ids')
+        answer_by_column, user_input_lines_ids = self._get_filters_from_post(post)
+
+        # Matrix, Multiple choice, Simple choice filters
+        if answer_by_column:
+            answer_ids, row_ids = [], []
+            for answer_column_id, answer_row_ids in answer_by_column.items():
+                answer_ids.append(answer_column_id)
+                row_ids += answer_row_ids
+
+            answers = request.env['survey.question.answer'].browse(list(answer_ids+row_ids))
+            # For performance, accessing 'a.matrix_question_id' caches all useful fields of the
+            # answers and rows records, avoiding unnecessary queries.
+            answers_only = answers.filtered(lambda a: not a.matrix_question_id)
```

Comment:
Hello @awa-odoo :slightly_smiling_face: 
My code has been updated :slightly_smiling_face: Thanks again! It simplifies everything a lot :slightly_smiling_face: 

About this comment, the extra queries come from the `row.value` so the access to the row fields. Each time we have a matrix filter, it adds a query as the row record hasn't been saved in cache and this `row.value` is in a loop on the rows.

The `answers.filtered(lambda a: not a.matrix_question_id)` is saving in cache fields from all the answers and the rows. So one query caching every needed records which maintains the query count at 3 whatever the number of matrix filters.

It's a bit difficult to explain sorry :sweat_smile: I hope it's a little bit more understandable now :slightly_smiling_face:
So I'm a bit lost here... I don't know if it can be seen as a side effect or not? Do I use the suggestion and increment the matrix query counts in the tests or do I leave it this way with a nice comment? There also might be another solution that I'm not aware of. :thinking: 

Waiting for your opinion! Thanks a lot :slightly_smiling_face: